desc/compute/__init__.py

```diff
@@ -52,16 +52,37 @@
 
 
 # rather than having to recursively compute the full dependencies every time we
-# compute something, its easier to just do it once for all quantities when we first
+# compute something, it's easier to just do it once for all quantities when we first
 # import the compute module.
 def _build_data_index():
     for key in data_index.keys():
-        full = {}
-        full["data"] = get_data_deps(key)
-        full["transforms"] = get_derivs(key)
-        full["params"] = get_params(key)
-        full["profiles"] = get_profiles(key)
+        full = {
+            "data": get_data_deps(key, False),
+            "transforms": get_derivs(key, False),
+            "params": get_params(key, False),
+            "profiles": get_profiles(key, False),
+        }
         data_index[key]["full_dependencies"] = full
 
+        full_with_axis_data = get_data_deps(key, True)
+        if len(full["data"]) >= len(full_with_axis_data):
+            # Then this quantity lacks dependencies that would only be required
+            # to evaluate this quantity's limit at the magnetic axis.
+            full_with_axis = full
+        else:
+            full_with_axis = {
+                "data": full_with_axis_data,
+                "transforms": get_derivs(key, True),
+                "params": get_params(key, True),
+                "profiles": get_profiles(key, True),
+            }
+            for _key, val in full_with_axis.items():
+                if _key == "transforms":
+                    continue
+                if len(full[_key]) >= len(val):
+                    # one is a deep copy of the other; dereference to save memory
+                    full_with_axis[_key] = full[_key]
+        data_index[key]["full_with_axis_dependencies"] = full_with_axis
```

Comment:
I feel like this is clunky, but I don't have a better idea at the moment.

Reply:
can you explain a bit what's going on here? Not sure I understand.

Reply:
Similar to [this comment](https://github.com/PlasmaControl/DESC/pull/448/files/ea78cdc354c0fc6fd3997a7ea789e77c91a7875b#r1222004632). Inside this `else` branch, we know that `key` requires an additional dependency to evaluate its limit at the axis. Typically this dependency will need the next higher order derivative of some quantity, so it will require an additional (or higher order) `transform`.  The evaluation of the limit will usually not need any additional profile like `current` or `iota`. `data["B0"]` is a good example of this.

In this loop of length 4, my code claims that when the length of the list returned by `get_params(key, has_axis=True)` is equal to the length of the list returned by `get_params(key, has_axis=False)`, then they must be duplicates of one another because no additional `params` are needed to evaluate the limit. Instead of keeping references to two copies of the same object, I just use the full_dependencies params reference. (This doesn't work on `transforms` because a list of transforms with equal length may have contents that differ.)

Reply:
I felt like it was clunky because we create two dictionaries whose internals are mostly identical, but we compute and store each them separately. The intention is to save memory by making the internals (depth level of 1 internals) point to the same instance when I believe the two things are copies.